docs: add Rule Extensions GA documentation (#1446)
snyk/user-docs: scan-fix-and-prevent/scan-with-snyk/snyk-code/rule-extensions/faq-and-troubleshooting.md

### Which languages are supported by rules extensions?

Rules extensions support most of the same languages as Snyk Code tests. The [Supported rules](supported-rules.md) page shows detailed support for combinations of rules, languages, and rule extensions. For more information on Snyk Code language support, refer to the [documentation](https://docs.snyk.io/supported-languages-package-managers-and-frameworks). For information about the roadmap, reach out to your account team.

## Migrating from the closed beta

If you used Rule Extensions during the closed beta, complete these steps for general availability:

* **Move to the GA API endpoints.** The closed-beta API endpoints will be retired 30 days after general availability. The GA endpoints are available now — see the [API reference](https://docs.snyk.io/developer-tools/snyk-api/reference/sastruleextensions).

> * **Update your custom-role permissions.** Grant your roles the `rule_extension.*` permissions they need, including **View Groups** and **View Organizations**. See [Required permissions](#required-permissions).

Reviewer: Broken link: this anchor doesn't exist on this page anymore — the permissions content moved to the new `rule-extensions-permissions.md` page. This link is currently dead.
```suggestion
* **Update your custom-role permissions.** Grant your roles the `rule_extension.*` permissions they need, including **View Groups** and **View Organizations**. See [Rule Extensions permissions](rule-extensions-permissions.md).
```

> ## Rule extensions

Reviewer: ```suggestion
## Rule Extensions
```

> ### How do I know if a rule extension is working, or the impact it is having?

Reviewer: nit: this page has many instances of lowercase "rule extension(s)" (e.g. lines 28, 30, 32, 34, 40, 42, 44, 46, 54, 58, 82, 86, 88, 90, 92, 98, 102, 104, 105, 200, 202, 219) where the rest of the PR treats "Rule Extension" as a capitalized proper noun/entity name. Worth a find-and-replace pass on this file for consistency rather than fixing each one individually here.

* **Non-public functions** — sanitizers must be public or exported.

See [Identify your sanitizer's FQN](identify-your-sanitizers-fqn.md) for per-language detail, and validate any uncertain case with [impact testing](impact-testing.md).

### How do I choose the right sanitizer type?

Match the type to how your function behaves — Flow Through (the return value is always clean), If True / If False (checked in a condition), or Any Usage (mutates in place or throws). Choosing **Any Usage** for a function that does not sanitize on every path can create **false negatives** by hiding a real vulnerability. When in doubt, run an impact test before publishing. See [Custom sanitizers](custom-sanitizers.md#types-of-sanitizers).

## Rule management

### What is the difference between `published` and `draft` for rule extensions?

* **Published** means the extension can be assigned and applied to Snyk Code scans (together with [assignments](https://docs.snyk.io/developer-tools/snyk-api/reference/assignments)).
* **Draft** keeps the configuration saved but it is not applied to scans until you publish it.

### Can published rule extensions be changed to draft?

> Published rule extensions cannot be changed to draft rule extensions. Create a new rule extension and delete the one that is no longer required.

Reviewer: ```suggestion
Published Rule Extensions cannot be changed to draft. Create a new Rule Extension and delete the one that's no longer needed.
```

### What data can I update for a rule?

The [Update a rule extension by rule extension ID](https://docs.snyk.io/developer-tools/snyk-api/reference/sastruleextensions#patch-groups-group_id-sast-rule_extensions-rule_extension_id) endpoint supports partial updates to `description`, `configuration` (sanitization `type` and `rule_keys`), and `status`. For **draft** extensions, you can also update `signature` (including `fully_qualified_name`). For **published** extensions, you **cannot** update `signature`.

### What do the different sanitizer types mean?

Sanitizer type refers to the way the input data is being sanitized. Snyk has identified four major types that are supported in Snyk scans, covering the different usage patterns. See [Custom sanitizers](custom-sanitizers.md).

### Can Snyk recover deleted rule extensions?

> Deleted rule extensions cannot be recovered. If a rule extension is deleted, you must create a new rule extension.

Reviewer: ```suggestion
Deleted Rule Extensions cannot be recovered. You must create a new one.
```

**Problematic code (reconstruction at call site)**

The string `addressLabelPath + pdf + ".pdf"` is constructed twice: once for the check and once for the delete. Snyk sees these as two different events.

{% code fullWidth="false" %}
```csharp
// The argument is reconstructed in the check
if (Sanitizer.PathIsSafe(addressLabelPath + pdf + ".pdf")) {
    // And reconstructed again in the sink
    System.IO.File.Delete(addressLabelPath + pdf + ".pdf");
}
```
{% endcode %}

**Recommended structure (assign to variable)**

Assign the constructed value to a variable first, then use that variable in both the sanitizer and the sink. This ensures Snyk tracks the same object.

{% code fullWidth="false" %}
```csharp
// Assign to a variable first
var fullPath = addressLabelPath + pdf + ".pdf";

// Use the variable in the check
if (Sanitizer.PathIsSafe(fullPath)) {
    // Use the same variable in the sink
    System.IO.File.Delete(fullPath);
}
```
{% endcode %}

## Upcoming plans and enhancements

### Will the API have Organization scope access in the future?

Snyk plans to revisit access and custom roles in the future.

> ### Will the rule extension scope be expanded to tenants in the future? That is, will we be able to create rule extensions across groups?

Reviewer: ```suggestion
### Will Rule Extensions be available across Groups (tenants) in the future?
```

## Troubleshooting

### Why is my code scan timing out?

Try running the test again, and if the issue persists, contact your account team or open a case with [Snyk support](https://support.snyk.io/s/).

### I accidentally deleted my rule; can I get it back?

Deleted rule extensions cannot be recovered. You must create a new one.

### I added a custom sanitizer but the false positives still show up; what should I do?

Snyk supports fully qualified function names or sanitizers that are uniquely identifiable across a repository or Project. If you are still having issues after ensuring your sanitizer meets these criteria, reach out to your account team or open a case with [Snyk support](https://support.snyk.io/s/).

For further assistance, contact your account team or refer to the [Snyk documentation](https://docs.snyk.io/).

## Glossary

| Term                  | Definition                                                                                                                     |
| --------------------- | ------------------------------------------------------------------------------------------------------------------------------ |

> | **FQN**               | Fully Qualified Name - the complete identifier for a function including namespace, class, and method name.                     |

Reviewer: ```suggestion
| **FQN**               | Fully Qualified Name — the complete identifier for a function including namespace, class, and method name.                     |
```

# FAQ & troubleshooting

This page provides answers to common questions about using Rule Extensions, along with troubleshooting guidance and permission requirements.

## General information

### What entitlements and permissions do I need to use this feature?

You must be an Enterprise customer with API entitlement and have the appropriate permissions configured through a custom role.

#### Required permissions

> Rule Extensions are **managed at the Group level** and **tested at the Organization level**. Access is granted through a [custom role](https://docs.snyk.io/snyk-platform-administration/user-roles/custom-role-templates): create or edit a role and add only the permissions for the actions that the role needs to perform.

Reviewer 2: Not entirely sure adding the custom-role-templates link would be helpful here. Unless we are planning to add an entry on custom roles for Rule Extensions?

You do **not** need every permission listed below. Grant only what the role requires:

* A read-only reviewer needs **View Rule Extensions**, **View Groups**, and **View Organizations**.
* A user who manages extensions also needs the **Create**, **Edit**, and/or **Delete** permissions.
* A user who runs impact tests needs the Organization-level **Test Rule Extensions** permission.

When you configure the custom role in the Snyk Web UI, the Group-level permissions appear under the **Rule Extensions Management** permission group.

> **Group-level permissions**

Reviewer 2: nit: should we have a specific page to outline the required permissions? And we can link this pages to here?

Reviewer 2: I have updated the https://snyksec.atlassian.net/wiki/spaces/coin/pages/4934238242/Creating+a+custom+role+for+Sast+or+Secrets+Rule+Extensions. Depending on the format how we access Rule Extensions, the required permissions for UI and API access are slightly different.

| View Groups            | `group.read`                  | View the Group that a rule extension belongs to. Required to manage Group-scoped rule extensions.                                 |
| View Organizations     | `group.org.list`              | See the Organizations a rule extension applies to. Required because rule extensions are scoped to Organizations within the Group. |

**Organization-level permission**

This permission is only needed to use the impact testing API, which runs a rule extension against a Project to preview its results.

| Permission name           | Permission key                          | Grants                                                                       |
| ------------------------- | --------------------------------------- | ----------------------------------------------------------------------------- |
| Test Rule Extensions | `org.rule_extension.project.test` | Run the impact test for a rule extension on a Project in the Organization. |

### What are the API rate limits?

> The rate limits are 1 request per second, 10 requests per minute, and 100 requests per hour.

Reviewer 2: This would be different now. The current API rate limit is for closed beta. 

For rest,  https://github.com/snyk/api-gateway/blob/main/helm/templates/rate-limits/api-rest.yaml
* Impact Test POST and GET results would be in the **low bucket**
* The rest of the API would be in the default **high bucket**

## Rule extensions

### How do I know if a rule extension is working, or the impact it is having?

The best way to understand the impact of a rule extension is by using the [impact testing API](https://docs.snyk.io/developer-tools/snyk-api/reference/sastruleextensions#post-groups-group_id-sast-rule_extensions-tests). Use the smallest practical project for the test.

Snyk will be introducing **UI** functionality that allows you to test rule extension changes to a Project for the General Availability milestone.

### What is the delay between publishing a rule extension and being able to see the results in a Code scan?

The maximum delay is nine (9) hours.

{% hint style="warning" %}
A delay of up to 9 hours can occur before a rule extension is recognized when Snyk scans an existing commit hash. To force the change to be picked up sooner, push a new commit to the repository.

> {% endhint %}

Reviewer 2: Do we still want to highlight this? Since we are actively trying to resolve it as the GA criteria?

Reviewer 3: Keeping this for now — the delay is being fixed as a GA criterion by end of month, and we'll update the docs once it lands.

### I accidentally deleted my rule; can I get it back?

Deleted rule extensions cannot be recovered. You must create a new one.

### I added a custom sanitizer but the false positives still show up; what should I do?

Snyk supports fully qualified function names or sanitizers that are uniquely identifiable across a repository or Project. If you are still having issues after ensuring your sanitizer meets these criteria, reach out to your account team or open a case with [Snyk support](https://support.snyk.io/s/).

For further assistance, contact your account team or refer to the [Snyk documentation](https://docs.snyk.io/).

## Glossary

| Term                  | Definition                                                                                                                     |
| --------------------- | ------------------------------------------------------------------------------------------------------------------------------ |
| **FQN**               | Fully Qualified Name - the complete identifier for a function including namespace, class, and method name.                     |

> | **Rule Name**         | Lists all rules that the extension applies to.                                                                                 |

Reviewer 2: Should this be `Rule Key` now, as we don't have this term in the API spec?
